---
 base/logging.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: src/base/logging.h
===================================================================
--- src.orig/base/logging.h	2012-11-07 10:19:52.000000000 +0000
+++ src/base/logging.h	2012-11-12 11:35:58.098865742 +0000
@@ -449,7 +449,7 @@
 // uses OFFICIAL_BUILD and other platforms use the branding flag when NDEBUG is
 // defined.
 #if ( defined(OS_WIN) && defined(OFFICIAL_BUILD)) || \
-    (!defined(OS_WIN) && defined(NDEBUG) && defined(GOOGLE_CHROME_BUILD))
+    (!defined(OS_WIN) && defined(NDEBUG))
 #define LOGGING_IS_OFFICIAL_BUILD 1
 #else
 #define LOGGING_IS_OFFICIAL_BUILD 0
