Index: src/chrome/browser/ui/browser.cc
===================================================================
--- src.orig/chrome/browser/ui/browser.cc	2013-03-25 16:11:50.144011578 +0100
+++ src/chrome/browser/ui/browser.cc	2013-03-25 16:11:50.164011579 +0100
@@ -329,6 +329,15 @@
 }
 
 // static
+Browser::CreateParams Browser::CreateParams::CreateForAppDesktop(
+    const std::string& app_name,
+    Profile* profile) {
+  CreateParams params(TYPE_DESKTOP, profile);
+  params.app_name = app_name;
+  return params;
+}
+
+// static
 Browser::CreateParams Browser::CreateParams::CreateForDevTools(
     Profile* profile) {
   CreateParams params(TYPE_POPUP, profile);
@@ -556,11 +565,6 @@
   return name;
 }
 
-GURL Browser::GetURLForCurrentTab() const {
-  WebContents* contents = chrome::GetActiveWebContents(this);
-  return contents ? contents->GetURL() : GURL();
-}
-
 ///////////////////////////////////////////////////////////////////////////////
 // Getters & Setters
 
@@ -601,7 +605,7 @@
 }
 
 GURL Browser::GetURLForCurrentTab() const {
-  WebContents* contents = GetActiveWebContents();
+  WebContents* contents = chrome::GetActiveWebContents(this);
   // |contents| can be NULL because GetWindowTitleForCurrentTab is called by the
   // window during the window's creation (before tabs have been added).
   return contents ? contents->GetURL() : GURL();
@@ -651,7 +655,9 @@
 // Browser, OnBeforeUnload handling:
 
 bool Browser::ShouldCloseWindow() {
-  if (!CanCloseWithInProgressDownloads())
+  if (!CanCloseWithInProgressDownloads()
+      || (is_type_desktop() && !unload_controller_->HasReceivedKillSignal())
+  )
     return false;
 
   return unload_controller_->ShouldCloseWindow();
@@ -2223,25 +2229,6 @@
       }
     }
   }
-
-  // We need to register the app-icon pref.
-  std::string icon_pref(prefs::kBrowserWindowIcon);
-  icon_pref.append("_");
-  icon_pref.append(app_name);
-  prefs->RegisterFilePathPref(icon_pref.c_str(), FilePath(),
-			      PrefService::UNSYNCABLE_PREF);
-}
-
-void Browser::RegisterAppDesktopPrefs(const std::string& app_name, Profile* profile) {
-  // We need to register the window position pref.
-  std::string window_pref(prefs::kBrowserWindowPlacement);
-  window_pref.append("_");
-  window_pref.append(app_name);
-  PrefService* prefs = profile->GetPrefs();
-  if (!prefs->FindPreference(window_pref.c_str())) {
-    prefs->RegisterDictionaryPref(window_pref.c_str(),
-                                  PrefService::UNSYNCABLE_PREF);
-  }
 }
 
 ///////////////////////////////////////////////////////////////////////////////
@@ -2345,6 +2332,14 @@
 
     if (!is_app())
       features |= FEATURE_LOCATIONBAR;
+
+    if (is_type_desktop()) {
+      features &= ~FEATURE_BOOKMARKBAR;
+      features &= ~FEATURE_TITLEBAR;
+      features &= ~FEATURE_TABSTRIP;
+      features &= ~FEATURE_TOOLBAR;
+      features &= ~FEATURE_LOCATIONBAR;
+    }
   }
   return !!(features & feature);
 }
Index: src/chrome/browser/ui/browser.h
===================================================================
--- src.orig/chrome/browser/ui/browser.h	2013-03-25 16:11:50.144011578 +0100
+++ src/chrome/browser/ui/browser.h	2013-03-25 16:11:50.164011579 +0100
@@ -118,7 +118,8 @@
     // BrowserTest.StartMaximized.
     TYPE_TABBED = 1,
     TYPE_POPUP = 2,
-    TYPE_PANEL = 3
+    TYPE_PANEL = 3,
+    TYPE_DESKTOP = 4
   };
 
   // Distinguishes between browsers that host an app (opened from
@@ -174,6 +175,11 @@
                                      const gfx::Rect& window_bounds,
                                      Profile* profile);
 
+    // Like Create, but creates a tabstrip-less and toolbar-less "app" window for
+    // the specified app, as the user's desktop.
+    static CreateParams CreateForAppDesktop(const std::string& app_name,
+                                            Profile* profile);
+
     static CreateParams CreateForDevTools(Profile* profile);
 
     // The browser type.
@@ -472,6 +478,7 @@
   bool is_type_tabbed() const { return type_ == TYPE_TABBED; }
   bool is_type_popup() const { return type_ == TYPE_POPUP; }
   bool is_type_panel() const { return type_ == TYPE_PANEL; }
+  bool is_type_desktop() const { return type_ == TYPE_DESKTOP; }
 
   bool is_app() const;
   bool is_devtools() const;
@@ -802,8 +809,6 @@
                          int index,
                          DetachType type);
 
-  static void RegisterAppDesktopPrefs(const std::string& app_name, Profile* profile);
-
   // Shared code between Reload() and ReloadIgnoringCache().
   void ReloadInternal(WindowOpenDisposition disposition, bool ignore_cache);
 
Index: src/chrome/browser/ui/gtk/browser_window_gtk.cc
===================================================================
--- src.orig/chrome/browser/ui/gtk/browser_window_gtk.cc	2013-03-25 16:11:49.280011539 +0100
+++ src/chrome/browser/ui/gtk/browser_window_gtk.cc	2013-03-27 17:14:47.556078100 +0100
@@ -871,6 +871,42 @@
   return fullscreen_exit_bubble_ != NULL;
 }
 
+void BrowserWindowGtk::ResizeToDesktopSize() {
+  GdkScreen* gdkScreen = gtk_window_get_screen(window_);
+  GdkRectangle rect;
+
+  gdk_screen_get_monitor_geometry(gdkScreen, 0, &rect);
+
+  // Decrease the desktop size by the size of the gnome-panel.
+  // FIXME: Figure this out properly!
+  rect.y = 24;
+  rect.height -= rect.y;
+
+  gtk_window_move( window_, rect.x, rect.y );
+  gtk_window_resize( window_, rect.width, rect.height );
+}
+
+gboolean OnVisibilityNotify(GtkWindow* window, GdkEventVisibility* event,
+                             BrowserWindowGtk* browser_win) {
+  /* If we are a desktop window, this expose event may have been
+     triggered by a change in the screen dimensions, so resize.
+  */
+  if (window == browser_win->window() && gtk_window_get_type_hint(window) == GDK_WINDOW_TYPE_HINT_DESKTOP) {
+    browser_win->ResizeToDesktopSize();
+    return TRUE;
+  }
+  return FALSE;
+}
+
+void BrowserWindowGtk::SetDesktop() {
+  ResizeToDesktopSize();
+  gtk_window_activate_focus( window_ );
+  gtk_window_set_type_hint( window_, GDK_WINDOW_TYPE_HINT_DESKTOP );
+  gtk_window_set_skip_taskbar_hint( window_, FALSE );
+  gtk_window_set_skip_pager_hint( window_, FALSE );
+  g_signal_connect(G_OBJECT(window_), "visibility-notify-event", G_CALLBACK(OnVisibilityNotify), this);
+}
+
 LocationBar* BrowserWindowGtk::GetLocationBar() const {
   return toolbar_->GetLocationBar();
 }
@@ -1852,6 +1888,12 @@
 
   if (chrome::ShouldSaveWindowPlacement(browser_.get()))
     chrome::SaveWindowPlacement(browser_.get(), restored_bounds_, show_state);
+  else {
+    // Do not continue with SaveWindowPosition() if the previous
+    // ShouldSaveWindowPlacement() returns false, such is the case when
+    // running in desktop mode: window preference settings will crash <!>
+    return;
+  }
 
   // We also need to save the placement for startup.
   // This is a web of calls between views and delegates on Windows, but the
Index: src/ui/base/x/x11_util.cc
===================================================================
--- src.orig/ui/base/x/x11_util.cc	2013-03-08 19:05:28.000000000 +0100
+++ src/ui/base/x/x11_util.cc	2013-03-25 16:11:50.164011579 +0100
@@ -917,6 +917,36 @@
 #endif
 }
 
+bool ChangeProperty(
+    XID window, const std::string& property_name, const std::string& set_name) {
+  Atom property_atom = gdk_x11_get_xatom_by_name_for_display(
+      gdk_display_get_default(), property_name.c_str());
+  Atom set_atom = gdk_x11_get_xatom_by_name_for_display(
+      gdk_display_get_default(), set_name.c_str());
+  Atom type = XA_ATOM;
+  int format = 32;
+  int mode = PropModeReplace;
+  long unsigned int num_items = 1;
+
+  LOG(INFO) << "Setting X desktop mode hint";
+
+  int result = XChangeProperty( GDK_DISPLAY_XDISPLAY( gdk_display_get_default() ),
+                                window,
+                                property_atom,
+                                type,
+                                format,
+                                mode,
+                                (guchar*) &set_atom,
+                                num_items);
+
+  LOG(INFO) << "XChangeProperty returned: " << result;
+
+  if (result != Success)
+    return false;
+
+  return true;
+}
+
 XID GetParentWindow(XID window) {
   XID root = None;
   XID parent = None;
Index: src/ui/base/x/x11_util.h
===================================================================
--- src.orig/ui/base/x/x11_util.h	2013-03-08 19:05:28.000000000 +0100
+++ src/ui/base/x/x11_util.h	2013-03-25 16:11:50.164011579 +0100
@@ -169,6 +169,8 @@
                                     std::vector<Atom>* value);
 UI_EXPORT bool GetStringProperty(
     XID window, const std::string& property_name, std::string* value);
+bool ChangeProperty(
+    XID window, const std::string& property_name, const std::string& set_name);
 
 UI_EXPORT bool SetIntProperty(XID window,
                               const std::string& name,
Index: src/chrome/browser/ui/browser_window.h
===================================================================
--- src.orig/chrome/browser/ui/browser_window.h	2013-03-08 19:01:55.000000000 +0100
+++ src/chrome/browser/ui/browser_window.h	2013-03-25 16:11:50.164011579 +0100
@@ -130,6 +130,9 @@
   virtual bool IsInMetroSnapMode() const = 0;
 #endif
 
+  // Sets the desktop-mode state
+  virtual void SetDesktop() = 0;
+
   // Returns the location bar.
   virtual LocationBar* GetLocationBar() const = 0;
 
Index: src/chrome/browser/ui/gtk/browser_window_gtk.h
===================================================================
--- src.orig/chrome/browser/ui/gtk/browser_window_gtk.h	2013-03-08 19:01:59.000000000 +0100
+++ src/chrome/browser/ui/gtk/browser_window_gtk.h	2013-03-25 16:11:50.164011579 +0100
@@ -107,6 +107,8 @@
       FullscreenExitBubbleType bubble_type) OVERRIDE;
   virtual bool IsFullscreen() const OVERRIDE;
   virtual bool IsFullscreenBubbleVisible() const OVERRIDE;
+  void ResizeToDesktopSize();
+  virtual void SetDesktop();
   virtual LocationBar* GetLocationBar() const OVERRIDE;
   virtual void SetFocusToLocationBar(bool select_all) OVERRIDE;
   virtual void UpdateReloadStopState(bool is_loading, bool force) OVERRIDE;
Index: src/chrome/common/chrome_switches.cc
===================================================================
--- src.orig/chrome/common/chrome_switches.cc	2013-03-25 16:11:50.144011578 +0100
+++ src/chrome/common/chrome_switches.cc	2013-03-25 16:11:50.164011579 +0100
@@ -245,6 +245,9 @@
 // device management backend.
 const char kDeviceManagementUrl[]           = "device-management-url";
 
+// Specifies that the browser should be launched in "X Desktop" mode.
+const char kDesktop[]                       = "desktop";
+
 // Triggers a plethora of diagnostic modes.
 const char kDiagnostics[]                   = "diagnostics";
 
Index: src/chrome/common/chrome_switches.h
===================================================================
--- src.orig/chrome/common/chrome_switches.h	2013-03-25 16:11:50.148011578 +0100
+++ src/chrome/common/chrome_switches.h	2013-03-25 16:11:50.164011579 +0100
@@ -76,6 +76,7 @@
 extern const char kDebugEnableFrameToggle[];
 extern const char kDebugPackedApps[];
 extern const char kDebugPrint[];
+extern const char kDesktop[];
 extern const char kDeviceManagementUrl[];
 extern const char kDiagnostics[];
 extern const char kDisableAsyncDns[];
Index: src/chrome/browser/ui/panels/panel.cc
===================================================================
--- src.orig/chrome/browser/ui/panels/panel.cc	2013-03-08 19:01:55.000000000 +0100
+++ src/chrome/browser/ui/panels/panel.cc	2013-03-25 16:11:50.168011579 +0100
@@ -520,6 +520,10 @@
   return false;
 }
 
+void Panel::SetDesktop() {
+  NOTIMPLEMENTED();
+}
+
 void Panel::OnContentsAutoResized(const gfx::Size& new_content_size) {
   DCHECK(auto_resizable_);
   if (!collection_)
Index: src/chrome/browser/ui/panels/panel.h
===================================================================
--- src.orig/chrome/browser/ui/panels/panel.h	2013-03-08 19:01:55.000000000 +0100
+++ src/chrome/browser/ui/panels/panel.h	2013-03-25 16:11:50.168011579 +0100
@@ -119,6 +119,7 @@
   virtual bool IsMaximized() const OVERRIDE;
   virtual bool IsMinimized() const OVERRIDE;
   virtual bool IsFullscreen() const OVERRIDE;
+  virtual void SetDesktop() OVERRIDE;
   virtual gfx::NativeWindow GetNativeWindow() OVERRIDE;
   virtual gfx::Rect GetRestoredBounds() const OVERRIDE;
   virtual gfx::Rect GetBounds() const OVERRIDE;
Index: src/chrome/browser/ui/cocoa/browser_window_cocoa.h
===================================================================
--- src.orig/chrome/browser/ui/cocoa/browser_window_cocoa.h	2013-03-08 19:02:17.000000000 +0100
+++ src/chrome/browser/ui/cocoa/browser_window_cocoa.h	2013-03-25 16:11:50.168011579 +0100
@@ -72,6 +72,7 @@
       FullscreenExitBubbleType bubble_type) OVERRIDE;
   virtual bool IsFullscreen() const OVERRIDE;
   virtual bool IsFullscreenBubbleVisible() const OVERRIDE;
+  virtual void SetDesktop();
   virtual LocationBar* GetLocationBar() const OVERRIDE;
   virtual void SetFocusToLocationBar(bool select_all) OVERRIDE;
   virtual void UpdateReloadStopState(bool is_loading, bool force) OVERRIDE;
Index: src/chrome/browser/ui/cocoa/browser_window_cocoa.mm
===================================================================
--- src.orig/chrome/browser/ui/cocoa/browser_window_cocoa.mm	2013-03-08 19:02:17.000000000 +0100
+++ src/chrome/browser/ui/cocoa/browser_window_cocoa.mm	2013-03-25 16:11:50.168011579 +0100
@@ -355,6 +355,10 @@
   return false;
 }
 
+void BrowserWindowCocoa::SetDesktop() {
+  NOTIMPLEMENTED();
+}
+
 void BrowserWindowCocoa::ConfirmAddSearchProvider(
     TemplateURL* template_url,
     Profile* profile) {
Index: src/chrome/browser/ui/extensions/application_launch.cc
===================================================================
--- src.orig/chrome/browser/ui/extensions/application_launch.cc	2013-03-25 16:11:50.148011578 +0100
+++ src/chrome/browser/ui/extensions/application_launch.cc	2013-03-27 17:42:27.500153931 +0100
@@ -72,6 +72,7 @@
     extension_misc::LaunchContainer container,
     const GURL& url_input,
     const FilePath& icon_id,
+    bool as_desktop,
     Browser** app_browser,
     const gfx::Rect& override_bounds) {
   DCHECK(!url_input.is_empty() || extension);
@@ -92,6 +93,9 @@
     window_bounds = override_bounds;
   }
 
+  if (as_desktop)
+    type = Browser::TYPE_DESKTOP;
+
   Browser::CreateParams params(type, profile, chrome::GetActiveDesktop());
   params.app_name = app_name;
   params.initial_bounds = window_bounds;
@@ -133,6 +137,9 @@
     profile->GetPrefs()->SetFilePath(icon_pref_key.c_str(), icon_id);
   }
 
+  if (as_desktop)
+    browser->window()->SetDesktop();
+
   WebContents* web_contents = chrome::AddSelectedTabWithURL(
       browser, url, content::PAGE_TRANSITION_AUTO_TOPLEVEL);
   web_contents->GetMutableRendererPrefs()->can_accept_load_drops = false;
@@ -275,7 +282,7 @@
     case extension_misc::LAUNCH_PANEL:
     case extension_misc::LAUNCH_WINDOW:
       tab = OpenApplicationWindow(profile, extension, container,
-                                  override_url, empty_icon, NULL, gfx::Rect());
+                                  override_url, empty_icon, false, NULL, gfx::Rect());
       break;
     case extension_misc::LAUNCH_TAB: {
       tab = OpenApplicationTab(profile, extension, override_url,
@@ -291,6 +298,7 @@
 
 WebContents* OpenAppShortcutWindow(Profile* profile,
 				   const FilePath& icon_id,
+				   bool as_desktop,
                                    const GURL& url,
                                    const gfx::Rect& override_bounds) {
   Browser* app_browser;
@@ -300,6 +308,7 @@
       extension_misc::LAUNCH_WINDOW,
       url,
       icon_id,
+      as_desktop,
       &app_browser,
       override_bounds);
 
Index: src/chrome/browser/ui/extensions/application_launch.h
===================================================================
--- src.orig/chrome/browser/ui/extensions/application_launch.h	2013-03-25 16:11:50.148011578 +0100
+++ src/chrome/browser/ui/extensions/application_launch.h	2013-03-25 16:11:50.168011579 +0100
@@ -68,6 +68,7 @@
 // application_launch::OpenApplication().
 content::WebContents* OpenAppShortcutWindow(Profile* profile,
 					    const FilePath& icon_id,
+					    bool as_desktop,
                                             const GURL& url,
                                             const gfx::Rect& override_bounds);
 
Index: src/chrome/browser/ui/startup/startup_browser_creator_impl.cc
===================================================================
--- src.orig/chrome/browser/ui/startup/startup_browser_creator_impl.cc	2013-03-25 16:11:50.148011578 +0100
+++ src/chrome/browser/ui/startup/startup_browser_creator_impl.cc	2013-03-25 16:11:50.168011579 +0100
@@ -461,6 +461,13 @@
   return (app_tab != NULL);
 }
 
+bool StartupBrowserCreatorImpl::IsDesktopLaunch() {
+  if (command_line_.HasSwitch(switches::kDesktop)) {
+    return true;
+  }
+  return false;
+}
+
 bool StartupBrowserCreatorImpl::OpenApplicationWindow(
     Profile* profile,
     content::WebContents** out_app_contents) {
@@ -534,6 +541,7 @@
       WebContents* app_tab = application_launch::OpenAppShortcutWindow(
           profile,
 	  FilePath(icon_id),
+	  IsDesktopLaunch(),
           url,
           override_bounds);
 
@@ -869,7 +877,8 @@
     chrome::ShowObsoleteOSPrompt(browser);
 
     if (browser_defaults::kOSSupportsOtherBrowsers &&
-        !command_line_.HasSwitch(switches::kNoDefaultBrowserCheck)) {
+        !command_line_.HasSwitch(switches::kNoDefaultBrowserCheck) &&
+	!IsDesktopLaunch()) {
       // Generally, the default browser prompt should not be shown on first
       // run. However, when the set-as-default dialog has been suppressed, we
       // need to allow it.
Index: src/chrome/browser/ui/startup/startup_browser_creator_impl.h
===================================================================
--- src.orig/chrome/browser/ui/startup/startup_browser_creator_impl.h	2013-03-08 19:02:12.000000000 +0100
+++ src/chrome/browser/ui/startup/startup_browser_creator_impl.h	2013-03-25 16:11:50.168011579 +0100
@@ -80,6 +80,9 @@
   // In this case |app_url| or |app_id| are populated if they're non-null.
   bool IsAppLaunch(Profile* profile, std::string* app_url, std::string* app_id);
 
+  // If the process was launched with the desktop flag, return true.
+  bool IsDesktopLaunch();
+
   // If IsAppLaunch is true, tries to open an application window.
   // If the app is specified to start in a tab, or IsAppLaunch is false,
   // returns false to specify default processing. |out_app_contents| is an
Index: src/chrome/browser/ui/browser_window_state.cc
===================================================================
--- src.orig/chrome/browser/ui/browser_window_state.cc	2013-03-08 19:02:17.000000000 +0100
+++ src/chrome/browser/ui/browser_window_state.cc	2013-03-25 16:11:50.168011579 +0100
@@ -60,6 +60,7 @@
     return browser_defaults::kRestorePopups || browser->is_devtools() ||
         browser->is_app();
   case Browser::TYPE_PANEL:
+  case Browser::TYPE_DESKTOP:
     // Do not save the window placement of panels.
     return false;
   default:
Index: src/chrome/browser/ui/unload_controller.cc
===================================================================
--- src.orig/chrome/browser/ui/unload_controller.cc	2013-03-08 19:01:55.000000000 +0100
+++ src/chrome/browser/ui/unload_controller.cc	2013-03-25 16:11:50.168011579 +0100
@@ -23,6 +23,7 @@
 UnloadController::UnloadController(Browser* browser)
     : browser_(browser),
       is_attempting_to_close_browser_(false),
+      kill_signal_(0),
       ALLOW_THIS_IN_INITIALIZER_LIST(weak_factory_(this)) {
   browser_->tab_strip_model()->AddObserver(this);
 }
@@ -36,6 +37,8 @@
   // that avoids the fast shutdown path where we just kill all the renderers.
   if (is_attempting_to_close_browser_)
     ClearUnloadState(contents, true);
+  if (browser_->is_type_desktop() && !kill_signal_)
+    return false;
   return !is_attempting_to_close_browser_;
 }
 
@@ -47,6 +50,10 @@
     return proceed;
   }
 
+  if (browser_->is_type_desktop() && !kill_signal_) {
+    proceed = false;
+  }
+
   if (!proceed) {
     CancelWindowClose();
     contents->SetClosedByUserGesture(false);
Index: src/chrome/browser/ui/unload_controller.h
===================================================================
--- src.orig/chrome/browser/ui/unload_controller.h	2013-03-08 19:02:17.000000000 +0100
+++ src/chrome/browser/ui/unload_controller.h	2013-03-25 16:11:50.168011579 +0100
@@ -44,6 +44,8 @@
   // its BeforeUnloadHandler.
   bool BeforeUnloadFired(content::WebContents* contents, bool proceed);
 
+  bool HasReceivedKillSignal() { return kill_signal_ != 0; }
+
   bool is_attempting_to_close_browser() const {
     return is_attempting_to_close_browser_;
   }
@@ -128,6 +130,8 @@
   // Browser window isn't just immediately closed.
   bool is_attempting_to_close_browser_;
 
+  int kill_signal_;
+
   base::WeakPtrFactory<UnloadController> weak_factory_;
 
   DISALLOW_COPY_AND_ASSIGN(UnloadController);
Index: src/chrome/browser/ui/views/frame/browser_view.cc
===================================================================
--- src.orig/chrome/browser/ui/views/frame/browser_view.cc	2013-03-08 19:02:09.000000000 +0100
+++ src/chrome/browser/ui/views/frame/browser_view.cc	2013-03-25 16:11:50.168011579 +0100
@@ -1064,6 +1064,10 @@
     immersive_mode_controller_->MaybeStackViewAtTop();
 }
 
+void BrowserView::SetDestop() {
+  NOTIMPLEMENTED();
+}
+
 LocationBar* BrowserView::GetLocationBar() const {
   return GetLocationBarView();
 }
Index: src/chrome/browser/ui/views/frame/browser_view.h
===================================================================
--- src.orig/chrome/browser/ui/views/frame/browser_view.h	2013-03-08 19:02:09.000000000 +0100
+++ src/chrome/browser/ui/views/frame/browser_view.h	2013-03-25 16:11:50.172011580 +0100
@@ -279,6 +279,7 @@
   virtual void SetMetroSnapMode(bool enable) OVERRIDE;
   virtual bool IsInMetroSnapMode() const OVERRIDE;
 #endif
+  virtual void SetDesktop() OVERRIDE;
   virtual LocationBar* GetLocationBar() const OVERRIDE;
   virtual void SetFocusToLocationBar(bool select_all) OVERRIDE;
   virtual void UpdateReloadStopState(bool is_loading, bool force) OVERRIDE;
