--- a/chrome/browser/ui/gtk/gtk_theme_provider.cc
+++ b/chrome/browser/ui/gtk/gtk_theme_provider.cc
@@ -286,6 +286,7 @@
   registrar_.Init(profile->GetPrefs());
   registrar_.Add(prefs::kUsesSystemTheme, this);
   use_gtk_ = profile->GetPrefs()->GetBoolean(prefs::kUsesSystemTheme);
+  use_gtk_buttons_ = false;
 
   BrowserThemeProvider::Init(profile);
 }
@@ -358,11 +359,12 @@
   if ((type == NotificationType::PREF_CHANGED) &&
       (*Details<std::string>(details).ptr() == prefs::kUsesSystemTheme))
     use_gtk_ = profile()->GetPrefs()->GetBoolean(prefs::kUsesSystemTheme);
+    use_gtk_buttons_ = false;
 }
 
 GtkWidget* GtkThemeProvider::BuildChromeButton() {
   GtkWidget* button = HoverControllerGtk::CreateChromeButton();
-  gtk_chrome_button_set_use_gtk_rendering(GTK_CHROME_BUTTON(button), use_gtk_);
+  gtk_chrome_button_set_use_gtk_rendering(GTK_CHROME_BUTTON(button), use_gtk_buttons_);
   chrome_buttons_.push_back(button);
 
   signals_->Connect(button, "destroy", G_CALLBACK(OnDestroyChromeButtonThunk),
@@ -386,6 +388,10 @@
   return use_gtk_;
 }
 
+bool GtkThemeProvider::UseGtkButtons() const {
+  return use_gtk_buttons_;
+}
+
 GdkColor GtkThemeProvider::GetGdkColor(int id) const {
   return gfx::SkColorToGdkColor(GetColor(id));
 }
@@ -632,7 +638,7 @@
   for (std::vector<GtkWidget*>::iterator it = chrome_buttons_.begin();
        it != chrome_buttons_.end(); ++it) {
     gtk_chrome_button_set_use_gtk_rendering(
-        GTK_CHROME_BUTTON(*it), use_gtk_);
+        GTK_CHROME_BUTTON(*it), use_gtk_buttons_);
   }
 
   GtkWindow* window = BrowserList::GetLastActive()->window()->GetNativeHandle();
--- a/chrome/browser/ui/gtk/gtk_theme_provider.h
+++ b/chrome/browser/ui/gtk/gtk_theme_provider.h
@@ -72,8 +72,9 @@
   // Creates a theme-aware vertical separator widget.
   GtkWidget* CreateToolbarSeparator();
 
-  // Whether we should use the GTK system theme.
+  // Whether we should use the GTK system theme and buttons.
   bool UseGtkTheme() const;
+  bool UseGtkButtons() const;
 
   // A wrapper around ui::ThemeProvider::GetColor, transforming the result to a
   // GdkColor.
@@ -243,8 +244,9 @@
   CHROMEGTK_CALLBACK_1(GtkThemeProvider, gboolean, OnSeparatorExpose,
                        GdkEventExpose*);
 
-  // Whether we should be using gtk rendering.
+  // Whether we should be using gtk rendering and buttons
   bool use_gtk_;
+  bool use_gtk_buttons_;
 
   // GtkWidgets that exist only so we can look at their properties (and take
   // their colors).
--- a/chrome/browser/ui/gtk/custom_button.cc
+++ b/chrome/browser/ui/gtk/custom_button.cc
@@ -341,5 +341,5 @@
 }
 
 bool CustomDrawButton::UseGtkTheme() {
-  return theme_provider_ && theme_provider_->UseGtkTheme();
+  return theme_provider_ && theme_provider_->UseGtkButtons();
 }
--- a/chrome/browser/ui/gtk/reload_button_gtk.cc
+++ b/chrome/browser/ui/gtk/reload_button_gtk.cc
@@ -194,7 +194,7 @@
 
 gboolean ReloadButtonGtk::OnExpose(GtkWidget* widget,
                                    GdkEventExpose* e) {
-  if (theme_provider_ && theme_provider_->UseGtkTheme())
+  if (theme_provider_ && theme_provider_->UseGtkButtons())
     return FALSE;
   return ((visible_mode_ == MODE_RELOAD) ? reload_ : stop_).OnExpose(
       widget, e, hover_controller_.GetCurrentValue());
@@ -222,9 +222,9 @@
 }
 
 void ReloadButtonGtk::UpdateThemeButtons() {
-  bool use_gtk = theme_provider_ && theme_provider_->UseGtkTheme();
+  bool use_gtk_buttons = theme_provider_ && theme_provider_->UseGtkButtons();
 
-  if (use_gtk) {
+  if (use_gtk_buttons) {
     gtk_widget_ensure_style(widget());
     GtkIconSet* icon_set = gtk_style_lookup_icon_set(
         widget()->style,
@@ -267,7 +267,7 @@
     gtk_widget_set_double_buffered(widget(), FALSE);
   }
 
-  gtk_chrome_button_set_use_gtk_rendering(GTK_CHROME_BUTTON(widget()), use_gtk);
+  gtk_chrome_button_set_use_gtk_rendering(GTK_CHROME_BUTTON(widget()), use_gtk_buttons);
 }
 
 void ReloadButtonGtk::OnDoubleClickTimer() {
