Index: src/chrome/common/chrome_switches.cc
===================================================================
--- src.orig/chrome/common/chrome_switches.cc	2010-06-05 12:32:02.000000000 +0200
+++ src/chrome/common/chrome_switches.cc	2010-06-23 15:35:19.062396607 +0200
@@ -429,6 +429,9 @@
 //                                            except for "www.google.com".
 const char kHostResolverRules[]             = "host-resolver-rules";
 
+// Specifies the icon_id in order to use the correct icon.
+const char kIconId[]                         = "icon-id";
+
 // Perform importing from another browser. The value associated with this
 // setting encodes the target browser and what items to import.
 const char kImport[]                        = "import";
Index: src/chrome/browser/gtk/gtk_util.cc
===================================================================
--- src.orig/chrome/browser/gtk/gtk_util.cc	2010-06-05 12:31:58.000000000 +0200
+++ src/chrome/browser/gtk/gtk_util.cc	2010-06-23 15:35:19.062396607 +0200
@@ -14,6 +14,7 @@
 #include "app/l10n_util.h"
 #include "app/resource_bundle.h"
 #include "app/x11_util.h"
+#include "base/command_line.h"
 #include "base/i18n/rtl.h"
 #include "base/linux_util.h"
 #include "base/logging.h"
@@ -21,6 +22,7 @@
 #include "chrome/browser/browser_window.h"
 #include "chrome/browser/gtk/cairo_cached_surface.h"
 #include "chrome/browser/gtk/gtk_theme_provider.h"
+#include "chrome/common/chrome_switches.h"
 #include "chrome/common/renderer_preferences.h"
 #include "grit/theme_resources.h"
 #include "third_party/skia/include/core/SkBitmap.h"
@@ -62,15 +64,6 @@
   return TRUE;
 }
 
-// Ownership of |icon_list| is passed to the caller.
-GList* GetIconList() {
-  ResourceBundle& rb = ResourceBundle::GetSharedInstance();
-  GList* icon_list = NULL;
-  icon_list = g_list_append(icon_list, rb.GetPixbufNamed(IDR_PRODUCT_ICON_32));
-  icon_list = g_list_append(icon_list, rb.GetPixbufNamed(IDR_PRODUCT_LOGO_16));
-  return icon_list;
-}
-
 // A process wide singleton that manages our usage of gdk
 // cursors. gdk_cursor_new() hits the disk in several places and GdkCursor
 // instances can be reused throughout the process.
@@ -141,6 +134,48 @@
 
 namespace gtk_util {
 
+// Ownership of |icon_list| is passed to the caller.
+GList* ProductIconList() {
+  ResourceBundle& rb = ResourceBundle::GetSharedInstance();
+  GList* icon_list = NULL;
+  icon_list = g_list_append(icon_list, rb.GetPixbufNamed(IDR_PRODUCT_ICON_32));
+  icon_list = g_list_append(icon_list, rb.GetPixbufNamed(IDR_PRODUCT_LOGO_16));
+  return icon_list;
+}
+
+GdkPixbuf* GetPixbufFromIconID(const std::string& icon_id) {
+  GdkPixbuf* icon_pixbuf = NULL;
+  GError* error = NULL;
+  GtkIconTheme* icon_theme = gtk_icon_theme_get_default();
+
+  if (!icon_id.empty() && gtk_icon_theme_has_icon(icon_theme, icon_id.c_str())) {
+    icon_pixbuf = gtk_icon_theme_load_icon(icon_theme, icon_id.c_str(), 64, GTK_ICON_LOOKUP_USE_BUILTIN, &error);
+    if (!icon_pixbuf) {
+      g_warning("Couldn't load icon: %s", error->message);
+      g_error_free(error);
+    }
+  }
+  return icon_pixbuf;
+}
+
+GList* GetIconList() {
+  GList* icon_list = NULL;
+
+  const CommandLine& parsed_command_line = *CommandLine::ForCurrentProcess();
+  const std::string& icon_id =
+      parsed_command_line.GetSwitchValueASCII(switches::kIconId);
+
+  GdkPixbuf* icon_pixbuf = GetPixbufFromIconID(icon_id);
+  if (icon_pixbuf) {
+    icon_list = g_list_append(icon_list, icon_pixbuf);
+  }
+  else {
+    icon_list = ProductIconList();
+  }
+
+  return icon_list;
+}
+
 GtkWidget* CreateLabeledControlsGroup(std::vector<GtkWidget*>* labels,
                                       const char* text, ...) {
   va_list ap;
@@ -519,13 +554,13 @@
 }
 
 void SetWindowIcon(GtkWindow* window) {
-  GList* icon_list = GetIconList();
+  GList* icon_list = gtk_util::GetIconList();
   gtk_window_set_icon_list(window, icon_list);
   g_list_free(icon_list);
 }
 
 void SetDefaultWindowIcon() {
-  GList* icon_list = GetIconList();
+  GList* icon_list = gtk_util::ProductIconList();
   gtk_window_set_default_icon_list(icon_list);
   g_list_free(icon_list);
 }
Index: src/chrome/common/chrome_switches.h
===================================================================
--- src.orig/chrome/common/chrome_switches.h	2010-06-05 12:32:02.000000000 +0200
+++ src/chrome/common/chrome_switches.h	2010-06-23 15:35:19.062396607 +0200
@@ -25,6 +25,7 @@
 extern const char kAlwaysEnableDevTools[];
 extern const char kApp[];
 extern const char kAppId[];
+extern const char kIconId[];
 extern const char kAppLaunchAsPanel[];
 extern const char kAppLauncherForNewTab[];
 extern const char kAuthServerWhitelist[];
Index: src/chrome/browser/browser.cc
===================================================================
--- src.orig/chrome/browser/browser.cc	2010-06-23 15:34:51.042391376 +0200
+++ src/chrome/browser/browser.cc	2010-06-23 15:35:19.072391170 +0200
@@ -11,6 +11,7 @@
 #include "app/l10n_util.h"
 #include "base/base_paths.h"
 #include "base/command_line.h"
+#include "base/file_path.h"
 #include "base/keyboard_codes.h"
 #include "base/logging.h"
 #include "base/path_service.h"
@@ -410,12 +411,21 @@
 }
 
 // static
-void Browser::OpenApplicationWindow(Profile* profile, const GURL& url,
+void Browser::OpenApplicationWindow(Profile* profile, const GURL& url, const FilePath& icon_id,
                                     bool as_panel) {
   std::wstring app_name = web_app::GenerateApplicationNameFromURL(url);
   RegisterAppPrefs(app_name);
 
   Browser* browser = Browser::CreateForApp(app_name, profile, as_panel);
+
+  // save icon-id value for later use by UpdateWindowIcon
+  if (!icon_id.empty()) {
+      std::wstring icon_pref_key = browser->GetWindowIconKey();
+      PrefService* prefs = g_browser_process->local_state();
+      DCHECK(prefs);
+      prefs->SetFilePath(icon_pref_key.c_str(), icon_id);
+  }
+
   browser->AddTabWithURL(url, GURL(), PageTransition::START_PAGE, true, -1,
                          false, NULL);
 
@@ -439,9 +449,11 @@
 }
 
 // static
+const FilePath::CharType kEmptyIcon[] = FILE_PATH_LITERAL("");
 void Browser::OpenApplicationWindow(Profile* profile, Extension* extension) {
+  FilePath empty_icon(kEmptyIcon);
   OpenApplicationWindow(profile,
-      extension->GetFullLaunchURL(),
+      extension->GetFullLaunchURL(), empty_icon,
       (extension->launch_container() == Extension::LAUNCH_PANEL));
 }
 
@@ -505,6 +517,15 @@
 ///////////////////////////////////////////////////////////////////////////////
 // Browser, State Storage and Retrieval for UI:
 
+std::wstring Browser::GetWindowIconKey() const {
+  std::wstring name(prefs::kBrowserWindowIcon);
+  if (!app_name_.empty()) {
+    name.append(L"_");
+    name.append(app_name_);
+  }
+  return name;
+}
+
 std::wstring Browser::GetWindowPlacementKey() const {
   std::wstring name(prefs::kBrowserWindowPlacement);
   if (!app_name_.empty()) {
@@ -577,6 +598,18 @@
   return contents ? contents->GetFavIcon() : SkBitmap();
 }
 
+GURL Browser::GetURLForCurrentTab() const {
+  TabContents* contents = tabstrip_model_.GetSelectedTabContents();
+  GURL url;
+
+  // |contents| can be NULL because GetWindowTitleForCurrentTab is called by the
+  // window during the window's creation (before tabs have been added).
+  if (contents)
+    url = contents->GetURL();
+
+  return url;
+}
+
 string16 Browser::GetWindowTitleForCurrentTab() const {
   TabContents* contents = tabstrip_model_.GetSelectedTabContents();
   string16 title;
@@ -1593,6 +1626,7 @@
 // static
 void Browser::RegisterPrefs(PrefService* prefs) {
   prefs->RegisterDictionaryPref(prefs::kBrowserWindowPlacement);
+  prefs->RegisterFilePathPref(prefs::kBrowserWindowIcon, FilePath());
   prefs->RegisterIntegerPref(prefs::kOptionsWindowLastTabIndex, 0);
   prefs->RegisterIntegerPref(prefs::kDevToolsSplitLocation, -1);
 }
@@ -3478,10 +3512,14 @@
   std::wstring window_pref(prefs::kBrowserWindowPlacement);
   window_pref.append(L"_");
   window_pref.append(app_name);
+  std::wstring icon_pref(prefs::kBrowserWindowIcon);
+  icon_pref.append(L"_");
+  icon_pref.append(app_name);
   PrefService* prefs = g_browser_process->local_state();
   DCHECK(prefs);
 
   prefs->RegisterDictionaryPref(window_pref.c_str());
+  prefs->RegisterFilePathPref(icon_pref.c_str(), FilePath());
 }
 
 // static
Index: src/chrome/browser/gtk/browser_titlebar.cc
===================================================================
--- src.orig/chrome/browser/gtk/browser_titlebar.cc	2010-06-23 15:34:51.432391021 +0200
+++ src/chrome/browser/gtk/browser_titlebar.cc	2010-06-23 15:36:10.604266885 +0200
@@ -13,11 +13,13 @@
 #include "app/l10n_util.h"
 #include "app/resource_bundle.h"
 #include "base/command_line.h"
+#include "base/logging.h"
 #include "base/singleton.h"
 #include "base/string_piece.h"
 #include "base/string_tokenizer.h"
 #include "chrome/app/chrome_dll_resource.h"
 #include "chrome/browser/browser.h"
+#include "chrome/browser/browser_process.h"
 #include "chrome/browser/encoding_menu_controller.h"
 #include "chrome/browser/gtk/accelerators_gtk.h"
 #include "chrome/browser/gtk/browser_window_gtk.h"
@@ -33,6 +35,7 @@
 #include "chrome/browser/pref_service.h"
 #include "chrome/browser/profile.h"
 #include "chrome/browser/tab_contents/tab_contents.h"
+#include "chrome/common/chrome_switches.h"
 #include "chrome/common/notification_service.h"
 #include "chrome/common/pref_names.h"
 #include "gfx/gtk_util.h"
@@ -437,14 +440,31 @@
     // Update the system app icon.  We don't need to update the icon in the top
     // left of the custom frame, that will get updated when the throbber is
     // updated.
-    SkBitmap icon = browser_window_->browser()->GetCurrentPageIcon();
-    if (icon.empty()) {
-      gtk_util::SetWindowIcon(window_);
-    } else {
-      GdkPixbuf* icon_pixbuf = gfx::GdkPixbufFromSkBitmap(&icon);
-      gtk_window_set_icon(window_, icon_pixbuf);
-      g_object_unref(icon_pixbuf);
+    // If we got an icon-id parameter (saved into prefs) and the current URL is a subpath of
+    // the app parameter use the icon-id icon
+
+    std::wstring current_host = UTF8ToWide(browser_window_->browser()->GetURLForCurrentTab().host());
+
+    GList* icon_list = NULL;
+    GdkPixbuf* icon_pixbuf = NULL;
+    std::wstring icon_pref_key = browser_window_->browser()->GetWindowIconKey();
+    size_t found = icon_pref_key.find(current_host);
+    PrefService* prefs = g_browser_process->local_state();
+    FilePath prefs_icon_id = prefs->GetFilePath(icon_pref_key.c_str());
+    if (!prefs_icon_id.empty())
+      icon_pixbuf = gtk_util::GetPixbufFromIconID(prefs_icon_id.value());
+    // The current URL is foreign, use the default app icons
+    if (found == std::string::npos) {
+      icon_list = gtk_util::ProductIconList();
+    }
+    else if (icon_pixbuf) {
+      icon_list = g_list_append(icon_list, icon_pixbuf);
+    }
+    else {
+      gtk_util::GetIconList();
     }
+    gtk_window_set_icon_list(window_, icon_list);
+    g_list_free(icon_list);
   }
 }
 
Index: src/chrome/browser/browser.h
===================================================================
--- src.orig/chrome/browser/browser.h	2010-06-05 12:32:00.000000000 +0200
+++ src/chrome/browser/browser.h	2010-06-23 15:35:19.072391170 +0200
@@ -202,7 +202,7 @@
   // is true, the application will be opened as a Browser::Type::APP_PANEL in
   // app panel window, otherwise it will be opened as a Browser::Type::APP,
   // a.k.a. "thin frame".
-  static void OpenApplicationWindow(Profile* profile, const GURL& url,
+  static void OpenApplicationWindow(Profile* profile, const GURL& url, const FilePath& icon_id,
                                     bool as_panel);
 
   // Open an application for |extension| in a new application window or panel.
@@ -230,6 +230,7 @@
   // State Storage and Retrieval for UI ///////////////////////////////////////
 
   // Save and restore the window position.
+  std::wstring GetWindowIconKey() const;
   std::wstring GetWindowPlacementKey() const;
   bool ShouldSaveWindowPlacement() const;
   void SaveWindowPlacement(const gfx::Rect& bounds, bool maximized);
@@ -239,6 +240,9 @@
   // Gets the FavIcon of the page in the selected tab.
   SkBitmap GetCurrentPageIcon() const;
 
+  // Gets the URL of the current tab.
+  GURL GetURLForCurrentTab() const;
+
   // Gets the title of the window based on the selected tab's title.
   string16 GetWindowTitleForCurrentTab() const;
 
Index: src/chrome/browser/gtk/gtk_util.h
===================================================================
--- src.orig/chrome/browser/gtk/gtk_util.h	2010-06-05 12:31:58.000000000 +0200
+++ src/chrome/browser/gtk/gtk_util.h	2010-06-23 15:35:19.072391170 +0200
@@ -51,6 +51,10 @@
 // Horizontal Spacing between controls in a form.
 const int kFormControlSpacing = 10;
 
+GdkPixbuf* GetPixbufFromIconID(const std::string& icon_id);
+GList* GetIconList();
+GList* ProductIconList();
+
 // Create a table of labeled controls, using proper spacing and alignment.
 // Arguments should be pairs of const char*, GtkWidget*, concluding with a
 // NULL.  The first argument is a vector in which to place all labels
Index: src/chrome/browser/browser_init.cc
===================================================================
--- src.orig/chrome/browser/browser_init.cc	2010-06-05 12:32:00.000000000 +0200
+++ src/chrome/browser/browser_init.cc	2010-06-23 15:35:19.072391170 +0200
@@ -9,6 +9,7 @@
 #include "base/event_recorder.h"
 #include "base/path_service.h"
 #include "base/sys_info.h"
+#include "base/logging.h"
 #include "chrome/browser/automation/automation_provider.h"
 #include "chrome/browser/automation/chrome_frame_automation_provider.h"
 #include "chrome/browser/browser_list.h"
@@ -594,7 +595,8 @@
         ChildProcessSecurityPolicy::GetInstance();
     if (policy->IsWebSafeScheme(url.scheme()) ||
         url.SchemeIs(chrome::kFileScheme)) {
-      Browser::OpenApplicationWindow(profile, url, false);
+        std::string icon_id(command_line_.GetSwitchValueASCII(switches::kIconId));
+        Browser::OpenApplicationWindow(profile, url, FilePath(icon_id), false);
       return true;
     }
   }
Index: src/chrome/common/pref_names.cc
===================================================================
--- src.orig/chrome/common/pref_names.cc	2010-06-05 12:32:02.000000000 +0200
+++ src/chrome/common/pref_names.cc	2010-06-23 15:35:19.072391170 +0200
@@ -555,6 +555,9 @@
 // window to restore on startup.
 const wchar_t kBrowserWindowPlacement[] = L"browser.window_placement";
 
+// Desktop Icon name
+const wchar_t kBrowserWindowIcon[] = L"browser.window_icon";
+
 // A collection of position, size, and other data relating to the task
 // manager window to restore on startup.
 const wchar_t kTaskManagerWindowPlacement[] = L"task_manager.window_placement";
Index: src/chrome/common/pref_names.h
===================================================================
--- src.orig/chrome/common/pref_names.h	2010-06-05 12:32:02.000000000 +0200
+++ src/chrome/common/pref_names.h	2010-06-23 15:35:19.072391170 +0200
@@ -195,6 +195,7 @@
 extern const wchar_t kUninstallLastObservedRunTimeSec[];
 
 extern const wchar_t kBrowserWindowPlacement[];
+extern const wchar_t kBrowserWindowIcon[];
 extern const wchar_t kTaskManagerWindowPlacement[];
 extern const wchar_t kPageInfoWindowPlacement[];
 extern const wchar_t kKeywordEditorWindowPlacement[];
