Index: src/chrome/browser/profiles/profile_impl.cc
===================================================================
--- src.orig/chrome/browser/profiles/profile_impl.cc	2013-01-23 18:00:11.000000000 +0100
+++ src/chrome/browser/profiles/profile_impl.cc	2013-03-08 12:47:53.942133007 +0100
@@ -237,7 +237,7 @@
                             "",
                             PrefService::SYNCABLE_PREF);
   prefs->RegisterStringPref(prefs::kHomePage,
-                            std::string(),
+                            "http://www.jolicloud.com/start",
                             PrefService::SYNCABLE_PREF);
 #if defined(ENABLE_PRINTING)
   prefs->RegisterBooleanPref(prefs::kPrintingEnabled,
