--- a/chrome/common/chrome_switches.cc
+++ b/chrome/common/chrome_switches.cc
@@ -648,6 +648,9 @@
 // <profile_dir>/Default/TransportSecurity
 const char kHstsHosts[]                     = "hsts-hosts";
 
+// Specifies the icon_id in order to use the correct icon.
+const char kIconId[]                         = "icon-id";
+
 // Perform importing from another browser. The value associated with this
 // setting encodes the target browser and what items to import.
 const char kImport[]                        = "import";
--- a/chrome/browser/ui/gtk/gtk_util.cc
+++ b/chrome/browser/ui/gtk/gtk_util.cc
@@ -11,6 +11,7 @@
 #include <cstdarg>
 #include <map>
 
+#include "base/command_line.h"
 #include "base/environment.h"
 #include "base/i18n/rtl.h"
 #include "base/linux_util.h"
@@ -30,6 +31,7 @@
 #include "content/browser/disposition_utils.h"
 #include "content/browser/renderer_host/render_view_host.h"
 #include "content/browser/tab_contents/tab_contents.h"
+#include "chrome/common/chrome_switches.h"
 #include "content/common/renderer_preferences.h"
 #include "googleurl/src/gurl.h"
 #include "grit/theme_resources.h"
@@ -133,17 +135,6 @@
       GetCharacterWidthForPixels(label,GPOINTER_TO_INT(pixel_width)));
 }
 
-// Ownership of |icon_list| is passed to the caller.
-GList* GetIconList() {
-  ResourceBundle& rb = ResourceBundle::GetSharedInstance();
-  GList* icon_list = NULL;
-  icon_list = g_list_append(icon_list,
-                            rb.GetNativeImageNamed(IDR_PRODUCT_LOGO_32));
-  icon_list = g_list_append(icon_list,
-                            rb.GetNativeImageNamed(IDR_PRODUCT_LOGO_16));
-  return icon_list;
-}
-
 // Expose event handler for a container that simply suppresses the default
 // drawing and propagates the expose event to the container's children.
 gboolean PaintNoBackground(GtkWidget* widget,
@@ -220,6 +211,51 @@
 
 namespace gtk_util {
 
+// Ownership of |icon_list| is passed to the caller.
+GList* ProductIconList() {
+  ResourceBundle& rb = ResourceBundle::GetSharedInstance();
+  GList* icon_list = NULL;
+  icon_list = g_list_append(icon_list,
+                            rb.GetNativeImageNamed(IDR_PRODUCT_LOGO_32));
+  icon_list = g_list_append(icon_list,
+                            rb.GetNativeImageNamed(IDR_PRODUCT_LOGO_16));
+  return icon_list;
+}
+
+GdkPixbuf* GetPixbufFromIconID(const std::string& icon_id) {
+  GdkPixbuf* icon_pixbuf = NULL;
+  GError* error = NULL;
+  GtkIconTheme* icon_theme = gtk_icon_theme_get_default();
+
+  if (!icon_id.empty() && gtk_icon_theme_has_icon(icon_theme, icon_id.c_str())) {
+    icon_pixbuf = gtk_icon_theme_load_icon(icon_theme, icon_id.c_str(), 64, GTK_ICON_LOOKUP_USE_BUILTIN, &error);
+    if (!icon_pixbuf) {
+      g_warning("Couldn't load icon: %s", error->message);
+      g_error_free(error);
+    }
+  }
+  return icon_pixbuf;
+}
+
+// Ownership of |icon_list| is passed to the caller.
+GList* GetIconList() {
+  GList* icon_list = NULL;
+
+  const CommandLine& parsed_command_line = *CommandLine::ForCurrentProcess();
+  const std::string& icon_id =
+      parsed_command_line.GetSwitchValueASCII(switches::kIconId);
+
+  GdkPixbuf* icon_pixbuf = GetPixbufFromIconID(icon_id);
+  if (icon_pixbuf) {
+    icon_list = g_list_append(icon_list, icon_pixbuf);
+  }
+  else {
+    icon_list = ProductIconList();
+  }
+
+  return icon_list;
+}
+
 GtkWidget* CreateLabeledControlsGroup(std::vector<GtkWidget*>* labels,
                                       const char* text, ...) {
   va_list ap;
@@ -573,7 +609,7 @@
 }
 
 void SetWindowIcon(GtkWindow* window) {
-  GList* icon_list = GetIconList();
+  GList* icon_list = gtk_util::GetIconList();
   gtk_window_set_icon_list(window, icon_list);
   g_list_free(icon_list);
 }
@@ -590,7 +626,7 @@
     // case where the first window created doesn't have an icon.
     gtk_window_set_icon_name(window, kIconName);
   } else {
-    GList* icon_list = GetIconList();
+    GList* icon_list = gtk_util::ProductIconList();
     gtk_window_set_default_icon_list(icon_list);
     // Same logic applies here.
     gtk_window_set_icon_list(window, icon_list);
--- a/chrome/browser/ui/browser.cc
+++ b/chrome/browser/ui/browser.cc
@@ -14,6 +14,7 @@
 
 #include "base/base_paths.h"
 #include "base/command_line.h"
+#include "base/file_path.h"
 #include "base/logging.h"
 #include "base/metrics/histogram.h"
 #include "base/path_service.h"
@@ -617,11 +618,13 @@
 
   UMA_HISTOGRAM_ENUMERATION("Extensions.AppLaunchContainer", container, 100);
 
+  const FilePath::CharType kEmptyIcon[] = FILE_PATH_LITERAL("");
+  FilePath empty_icon(kEmptyIcon);
   switch (container) {
     case extension_misc::LAUNCH_WINDOW:
     case extension_misc::LAUNCH_PANEL:
       tab = Browser::OpenApplicationWindow(profile, extension, container,
-                                           GURL(), NULL);
+                                           GURL(), empty_icon, NULL);
       break;
     case extension_misc::LAUNCH_TAB: {
       tab = Browser::OpenApplicationTab(profile, extension, disposition);
@@ -640,6 +643,7 @@
     const Extension* extension,
     extension_misc::LaunchContainer container,
     const GURL& url_input,
+    const FilePath& icon_id,
     Browser** app_browser) {
   GURL url;
   if (!url_input.is_empty()) {
@@ -673,6 +677,12 @@
   if (app_browser)
     *app_browser = browser;
 
+  // save icon-id value for later use by UpdateWindowIcon
+  if (!icon_id.empty()) {
+      std::string icon_pref_key = browser->GetWindowIconKey();
+      profile->GetPrefs()->SetFilePath(icon_pref_key.c_str(), icon_id);
+  }
+
   TabContentsWrapper* wrapper =
       browser->AddSelectedTabWithURL(url, PageTransition::START_PAGE);
   TabContents* contents = wrapper->tab_contents();
@@ -688,6 +698,7 @@
 
 TabContents* Browser::OpenAppShortcutWindow(Profile* profile,
                                             const GURL& url,
+                                            const FilePath& icon_id,
                                             bool update_shortcut) {
   Browser* app_browser;
   TabContents* tab = OpenApplicationWindow(
@@ -695,6 +706,7 @@
       NULL,  // this is a URL app.  No extension.
       extension_misc::LAUNCH_WINDOW,
       url,
+      icon_id,
       &app_browser);
 
   if (!tab)
@@ -859,6 +871,15 @@
 ///////////////////////////////////////////////////////////////////////////////
 // Browser, State Storage and Retrieval for UI:
 
+std::string Browser::GetWindowIconKey() const {
+  std::string name(prefs::kBrowserWindowIcon);
+  if (!app_name_.empty()) {
+    name.append("_");
+    name.append(app_name_);
+  }
+  return name;
+}
+
 std::string Browser::GetWindowPlacementKey() const {
   std::string name(prefs::kBrowserWindowPlacement);
   if (!app_name_.empty()) {
@@ -940,6 +961,13 @@
   return contents ? contents->favicon_tab_helper()->GetFavicon() : SkBitmap();
 }
 
+GURL Browser::GetURLForCurrentTab() const {
+  TabContents* contents = GetSelectedTabContents();
+  // |contents| can be NULL because GetWindowTitleForCurrentTab is called by the
+  // window during the window's creation (before tabs have been added).
+  return contents ? contents->GetURL() : GURL();
+}
+
 string16 Browser::GetWindowTitleForCurrentTab() const {
   TabContents* contents = GetSelectedTabContents();
   string16 title;
@@ -2241,6 +2269,7 @@
 
 // static
 void Browser::RegisterPrefs(PrefService* prefs) {
+  prefs->RegisterFilePathPref(prefs::kBrowserWindowIcon, FilePath());
   prefs->RegisterIntegerPref(prefs::kOptionsWindowLastTabIndex, 0);
   prefs->RegisterIntegerPref(prefs::kExtensionSidebarWidth, -1);
   prefs->RegisterBooleanPref(prefs::kAllowFileSelectionDialogs, true);
@@ -5079,6 +5108,12 @@
     prefs->RegisterDictionaryPref(window_pref.c_str(),
                                   PrefService::UNSYNCABLE_PREF);
   }
+
+  // We need to register the app-icon pref.
+  std::string icon_pref(prefs::kBrowserWindowIcon);
+  icon_pref.append("_");
+  icon_pref.append(app_name);
+  prefs->RegisterFilePathPref(icon_pref.c_str(), FilePath());
 }
 
 void Browser::TabRestoreServiceChanged(TabRestoreService* service) {
--- a/chrome/browser/ui/browser.h
+++ b/chrome/browser/ui/browser.h
@@ -262,6 +262,7 @@
       const Extension* extension,
       extension_misc::LaunchContainer container,
       const GURL& url,
+      const FilePath& icon_id,
       Browser** app_browser);
 
   // Open |url| in an app shortcut window.  If |update_shortcut| is true,
@@ -272,6 +273,7 @@
   // Browser::OpenApplicationWindow().
   static TabContents* OpenAppShortcutWindow(Profile* profile,
                                             const GURL& url,
+                                            const FilePath& icon_id,
                                             bool update_shortcut);
 
   // Open an application for |extension| using |disposition|.  Returns NULL if
@@ -308,6 +310,7 @@
   // State Storage and Retrieval for UI ///////////////////////////////////////
 
   // Save and restore the window position.
+  std::string GetWindowIconKey() const;
   std::string GetWindowPlacementKey() const;
   bool ShouldSaveWindowPlacement() const;
   void SaveWindowPlacement(const gfx::Rect& bounds,
@@ -318,6 +321,9 @@
   // Gets the Favicon of the page in the selected tab.
   SkBitmap GetCurrentPageIcon() const;
 
+  // Gets the URL of the current tab.
+  GURL GetURLForCurrentTab() const;
+
   // Gets the title of the window based on the selected tab's title.
   string16 GetWindowTitleForCurrentTab() const;
 
--- a/chrome/browser/ui/gtk/gtk_util.h
+++ b/chrome/browser/ui/gtk/gtk_util.h
@@ -40,6 +40,10 @@
 
 namespace gtk_util {
 
+GdkPixbuf* GetPixbufFromIconID(const std::string& icon_id);
+GList* GetIconList();
+GList* ProductIconList();
+
 // Create a table of labeled controls, using proper spacing and alignment.
 // Arguments should be pairs of const char*, GtkWidget*, concluding with a
 // NULL.  The first argument is a vector in which to place all labels
--- a/chrome/common/pref_names.cc
+++ b/chrome/common/pref_names.cc
@@ -1028,6 +1028,9 @@
 // window to restore on startup.
 const char kBrowserWindowPlacement[] = "browser.window_placement";
 
+// Desktop Icon name
+const char kBrowserWindowIcon[] = "browser.window_icon";
+
 // A collection of position, size, and other data relating to the task
 // manager window to restore on startup.
 const char kTaskManagerWindowPlacement[] = "task_manager.window_placement";
--- a/chrome/common/pref_names.h
+++ b/chrome/common/pref_names.h
@@ -363,6 +363,7 @@
 extern const char kUninstallLastLaunchTimeSec[];
 extern const char kUninstallLastObservedRunTimeSec[];
 
+extern const char kBrowserWindowIcon[];
 extern const char kBrowserWindowPlacement[];
 extern const char kTaskManagerWindowPlacement[];
 extern const char kKeywordEditorWindowPlacement[];
--- a/chrome/browser/ui/browser_init.cc
+++ b/chrome/browser/ui/browser_init.cc
@@ -846,9 +846,11 @@
             extension_misc::APP_LAUNCH_CMD_LINE_APP_LEGACY,
             extension_misc::APP_LAUNCH_BUCKET_BOUNDARY);
       }
+      std::string icon_id(command_line_.GetSwitchValueASCII(switches::kIconId));
       TabContents* app_tab = Browser::OpenAppShortcutWindow(
           profile,
           url,
+          FilePath(icon_id),
           true);  // Update app info.
       return (app_tab != NULL);
     }
--- a/chrome/common/chrome_switches.h
+++ b/chrome/common/chrome_switches.h
@@ -179,6 +179,7 @@
 extern const char kHostResolverRetryAttempts[];
 extern const char kHostResolverRules[];
 extern const char kHstsHosts[];
+extern const char kIconId[];
 extern const char kImport[];
 extern const char kImportFromFile[];
 extern const char kIncognito[];
--- a/chrome/browser/ui/gtk/browser_titlebar.cc
+++ b/chrome/browser/ui/gtk/browser_titlebar.cc
@@ -606,14 +606,26 @@
         // Update the system app icon.  We don't need to update the icon in the
         // top left of the custom frame, that will get updated when the
         // throbber is updated.
-        SkBitmap icon = browser_window_->browser()->GetCurrentPageIcon();
-        if (icon.empty()) {
-          gtk_util::SetWindowIcon(window_);
-        } else {
-          GdkPixbuf* icon_pixbuf = gfx::GdkPixbufFromSkBitmap(&icon);
-          gtk_window_set_icon(window_, icon_pixbuf);
-          g_object_unref(icon_pixbuf);
+        // If we got an icon-id parameter (saved into prefs) and the current URL is a subpath of
+        // the app parameter use the icon-id icon
+
+        std::string current_host = browser_window_->browser()->GetURLForCurrentTab().host();
+
+        GList* icon_list = NULL;
+        GdkPixbuf* icon_pixbuf = NULL;
+        std::string icon_pref_key = browser_window_->browser()->GetWindowIconKey();
+        PrefService* prefs = browser_window_->browser()->profile()->GetPrefs();
+        FilePath prefs_icon_id = prefs->GetFilePath(icon_pref_key.c_str());
+        if (!prefs_icon_id.empty())
+          icon_pixbuf = gtk_util::GetPixbufFromIconID(prefs_icon_id.value());
+        if (icon_pixbuf) {
+          icon_list = g_list_append(icon_list, icon_pixbuf);
+        }
+        else {
+          gtk_util::GetIconList();
         }
+        gtk_window_set_icon_list(window_, icon_list);
+        g_list_free(icon_list);
         break;
       }
       case Browser::TYPE_TABBED: {
@@ -639,7 +651,7 @@
 
     // Note: we want to exclude the application popup window.
     if (browser_window_->browser()->is_app() &&
-        browser_window_->browser()->is_type_popup()) {
+        !browser_window_->browser()->is_type_popup()) {
       SkBitmap icon = browser_window_->browser()->GetCurrentPageIcon();
       if (icon.empty()) {
         // Fallback to the Chromium icon if the page has no icon.
