Index: src/chrome/browser/browser.cc
===================================================================
--- src.orig/chrome/browser/browser.cc	2010-08-13 23:11:59.000000000 +0200
+++ src/chrome/browser/browser.cc	2010-08-16 15:15:39.643641141 +0200
@@ -1866,8 +1866,7 @@
   if (home_page_pref &&
       !home_page_pref->IsManaged() &&
       !prefs->HasPrefPath(prefs::kHomePage)) {
-    prefs->SetString(prefs::kHomePage,
-        GoogleURLTracker::kDefaultGoogleHomepage);
+    prefs->SetString(prefs::kHomePage, "http://www.jolicloud.com/start");
   }
   const PrefService::Preference* home_page_is_new_tab_page_pref =
       prefs->FindPreference(prefs::kHomePageIsNewTabPage);
