Index: src/chrome/browser/browser_main.cc
===================================================================
--- src.orig/chrome/browser/browser_main.cc	2010-05-05 15:57:17.834271592 +0200
+++ src/chrome/browser/browser_main.cc	2010-05-05 16:36:33.264266723 +0200
@@ -829,11 +829,10 @@
                                             FilePath(), &master_prefs);
     AddFirstRunNewTabs(&browser_init, master_prefs.new_tabs);
 
-    // If we are running in App mode, we do not want to show the importer
+    // Even if we are running in App mode, we do want to show the importer
     // (first run) UI.
     if (!first_run_ui_bypass &&
-        (parsed_command_line.HasSwitch(switches::kApp) ||
-         parsed_command_line.HasSwitch(switches::kNoFirstRun)))
+         parsed_command_line.HasSwitch(switches::kNoFirstRun))
       first_run_ui_bypass = true;
   }
 
