Index: src/chrome/common/chrome_paths_linux.cc
===================================================================
--- src.orig/chrome/common/chrome_paths_linux.cc
+++ src/chrome/common/chrome_paths_linux.cc
@@ -21,11 +21,7 @@
   scoped_ptr<base::Environment> env(base::Environment::Create());
   FilePath config_dir(
       base::nix::GetXDGDirectory(env.get(), "XDG_CONFIG_HOME", ".config"));
-#if defined(GOOGLE_CHROME_BUILD)
   *result = config_dir.Append("google-chrome");
-#else
-  *result = config_dir.Append("chromium");
-#endif
   return true;
 }
 
@@ -59,11 +55,7 @@
   scoped_ptr<base::Environment> env(base::Environment::Create());
   FilePath config_dir(
       base::nix::GetXDGDirectory(env.get(), "XDG_CONFIG_HOME", ".config"));
-#if defined(GOOGLE_CHROME_BUILD)
   *result = config_dir.Append("google-chrome-frame");
-#else
-  *result = config_dir.Append("chrome-frame");
-#endif
   return true;
 }
 
