--- a/chrome/app/resources/locale_settings_linux.grd
+++ b/chrome/app/resources/locale_settings_linux.grd
@@ -148,7 +148,7 @@
     <messages fallback_to_english="true">
       <!-- The font used in Web UI (e.g. History). -->
       <message name="IDS_WEB_FONT_FAMILY" use_name_for_id="true">
-        Arial, sans-serif
+        Liberation Sans, sans-serif
       </message>
 
       <!-- The relative font size in % used in Web UI (e.g. History).  -->
@@ -158,22 +158,22 @@
 
       <!-- The default value for |WebPreference::standard_font_family| -->
       <message name="IDS_STANDARD_FONT_FAMILY" use_name_for_id="true">
-        Times New Roman
+        Liberation Serif
       </message>
 
       <!-- The default value for |WebPreference::fixed_font_family| -->
       <message name="IDS_FIXED_FONT_FAMILY" use_name_for_id="true">
-        Courier New
+        Liberation Mono
       </message>
 
       <!-- The default value for |WebPreference::serif_font_family| -->
       <message name="IDS_SERIF_FONT_FAMILY" use_name_for_id="true">
-        Times New Roman
+        Liberation Serif
       </message>
 
       <!-- The default value for |WebPreference::sans_serif_font_family| -->
       <message name="IDS_SANS_SERIF_FONT_FAMILY" use_name_for_id="true">
-        Arial
+        Liberation Sans
       </message>
 
       <!-- The default value for |WebPreference::cursive_font_family| -->
