--- a/chrome/common/chrome_switches.cc
+++ b/chrome/common/chrome_switches.cc
@@ -785,6 +785,9 @@
 // Ignores GPU blacklist.
 const char kIgnoreGpuBlacklist[]            = "ignore-gpu-blacklist";
 
+// Specifies the icon_id in order to use the correct icon.
+const char kIconId[]                        = "icon-id";
+
 // Performs importing from another browser. The value associated with this
 // setting encodes the target browser and what items to import.
 const char kImport[]                        = "import";
--- a/chrome/browser/ui/gtk/gtk_util.cc
+++ b/chrome/browser/ui/gtk/gtk_util.cc
@@ -10,6 +10,7 @@
 #include <cstdarg>
 #include <map>
 
+#include "base/command_line.h"
 #include "base/environment.h"
 #include "base/i18n/rtl.h"
 #include "base/logging.h"
@@ -31,6 +32,7 @@
 #include "chrome/browser/ui/gtk/gtk_theme_service.h"
 #include "content/public/browser/render_view_host.h"
 #include "content/public/browser/web_contents.h"
+#include "chrome/common/chrome_switches.h"
 #include "googleurl/src/gurl.h"
 #include "grit/theme_resources.h"
 #include "grit/theme_resources_standard.h"
@@ -127,17 +129,6 @@
       GetCharacterWidthForPixels(label, GPOINTER_TO_INT(pixel_width)));
 }
 
-// Ownership of |icon_list| is passed to the caller.
-GList* GetIconList() {
-  ui::ResourceBundle& rb = ui::ResourceBundle::GetSharedInstance();
-  GList* icon_list = NULL;
-  icon_list = g_list_append(icon_list,
-      rb.GetNativeImageNamed(IDR_PRODUCT_LOGO_32).ToGdkPixbuf());
-  icon_list = g_list_append(icon_list,
-      rb.GetNativeImageNamed(IDR_PRODUCT_LOGO_16).ToGdkPixbuf());
-  return icon_list;
-}
-
 // Returns the avatar icon for |profile|.
 //
 // Returns NULL if there is only one profile; always returns an icon for
@@ -293,6 +284,51 @@
 
 namespace gtk_util {
 
+// Ownership of |icon_list| is passed to the caller.
+GList* ProductIconList() {
+  ResourceBundle& rb = ResourceBundle::GetSharedInstance();
+  GList* icon_list = NULL;
+  icon_list = g_list_append(icon_list,
+                            rb.GetNativeImageNamed(IDR_PRODUCT_LOGO_32));
+  icon_list = g_list_append(icon_list,
+                            rb.GetNativeImageNamed(IDR_PRODUCT_LOGO_16));
+  return icon_list;
+}
+
+GdkPixbuf* GetPixbufFromIconID(const std::string& icon_id) {
+  GdkPixbuf* icon_pixbuf = NULL;
+  GError* error = NULL;
+  GtkIconTheme* icon_theme = gtk_icon_theme_get_default();
+
+  if (!icon_id.empty() && gtk_icon_theme_has_icon(icon_theme, icon_id.c_str())) {
+    icon_pixbuf = gtk_icon_theme_load_icon(icon_theme, icon_id.c_str(), 64, GTK_ICON_LOOKUP_USE_BUILTIN, &error);
+    if (!icon_pixbuf) {
+      g_warning("Couldn't load icon: %s", error->message);
+      g_error_free(error);
+    }
+  }
+  return icon_pixbuf;
+}
+
+// Ownership of |icon_list| is passed to the caller.
+GList* GetIconList() {
+  GList* icon_list = NULL;
+
+  const CommandLine& parsed_command_line = *CommandLine::ForCurrentProcess();
+  const std::string& icon_id =
+      parsed_command_line.GetSwitchValueASCII(switches::kIconId);
+
+  GdkPixbuf* icon_pixbuf = GetPixbufFromIconID(icon_id);
+  if (icon_pixbuf) {
+    icon_list = g_list_append(icon_list, icon_pixbuf);
+  }
+  else {
+    icon_list = ProductIconList();
+  }
+
+  return icon_list;
+}
+
 GtkWidget* CreateLabeledControlsGroup(std::vector<GtkWidget*>* labels,
                                       const char* text, ...) {
   va_list ap;
@@ -584,7 +620,7 @@
     // case where the first window created doesn't have an icon.
     gtk_window_set_icon_name(window, kIconName);
   } else {
-    GList* icon_list = GetIconList();
+    GList* icon_list = gtk_util::ProductIconList();
     gtk_window_set_default_icon_list(icon_list);
     // Same logic applies here.
     gtk_window_set_icon_list(window, icon_list);
@@ -593,7 +629,7 @@
 }
 
 void SetWindowIcon(GtkWindow* window, Profile* profile) {
-  GList* icon_list = GetIconListWithAvatars(window, profile);
+  GList* icon_list = gtk_util::GetIconList();
   gtk_window_set_icon_list(window, icon_list);
   g_list_foreach(icon_list, reinterpret_cast<GFunc>(g_object_unref), NULL);
   g_list_free(icon_list);
--- a/chrome/browser/ui/browser.cc
+++ b/chrome/browser/ui/browser.cc
@@ -15,6 +15,7 @@
 #include "base/base_paths.h"
 #include "base/bind.h"
 #include "base/command_line.h"
+#include "base/file_path.h"
 #include "base/logging.h"
 #include "base/metrics/field_trial.h"
 #include "base/metrics/histogram.h"
@@ -738,6 +739,15 @@
 ///////////////////////////////////////////////////////////////////////////////
 // Browser, State Storage and Retrieval for UI:
 
+std::string Browser::GetWindowIconKey() const {
+  std::string name(prefs::kBrowserWindowIcon);
+  if (!app_name_.empty()) {
+    name.append("_");
+    name.append(app_name_);
+  }
+  return name;
+}
+
 std::string Browser::GetWindowPlacementKey() const {
   std::string name(prefs::kBrowserWindowPlacement);
   if (!app_name_.empty()) {
@@ -849,6 +859,13 @@
   return contents ? contents->favicon_tab_helper()->GetFavicon() : SkBitmap();
 }
 
+GURL Browser::GetURLForCurrentTab() const {
+  WebContents* contents = GetActiveWebContents();
+  // |contents| can be NULL because GetWindowTitleForCurrentTab is called by the
+  // window during the window's creation (before tabs have been added).
+  return contents ? contents->GetURL() : GURL();
+}
+
 string16 Browser::GetWindowTitleForCurrentTab() const {
   WebContents* contents = GetActiveWebContents();
   string16 title;
@@ -2070,6 +2087,7 @@
 
 // static
 void Browser::RegisterPrefs(PrefService* prefs) {
+  prefs->RegisterFilePathPref(prefs::kBrowserWindowIcon, FilePath());
   prefs->RegisterIntegerPref(prefs::kOptionsWindowLastTabIndex, 0);
   prefs->RegisterBooleanPref(prefs::kAllowFileSelectionDialogs, true);
   prefs->RegisterBooleanPref(prefs::kShouldShowFirstRunBubble, false);
@@ -4885,6 +4903,12 @@
     prefs->RegisterDictionaryPref(window_pref.c_str(),
                                   PrefService::UNSYNCABLE_PREF);
   }
+
+  // We need to register the app-icon pref.
+  std::string icon_pref(prefs::kBrowserWindowIcon);
+  icon_pref.append("_");
+  icon_pref.append(app_name);
+  prefs->RegisterFilePathPref(icon_pref.c_str(), FilePath());
 }
 
 void Browser::ReloadInternal(WindowOpenDisposition disposition,
--- a/chrome/browser/ui/browser.h
+++ b/chrome/browser/ui/browser.h
@@ -342,6 +342,7 @@
   // State Storage and Retrieval for UI ///////////////////////////////////////
 
   // Save and restore the window position.
+  std::string GetWindowIconKey() const;
   std::string GetWindowPlacementKey() const;
   bool ShouldSaveWindowPlacement() const;
   void SaveWindowPlacement(const gfx::Rect& bounds,
@@ -352,6 +353,9 @@
   // Gets the Favicon of the page in the selected tab.
   SkBitmap GetCurrentPageIcon() const;
 
+  // Gets the URL of the current tab.
+  GURL GetURLForCurrentTab() const;
+
   // Gets the title of the window based on the selected tab's title.
   string16 GetWindowTitleForCurrentTab() const;
 
--- a/chrome/browser/ui/gtk/gtk_util.h
+++ b/chrome/browser/ui/gtk/gtk_util.h
@@ -31,6 +31,10 @@
 
 namespace gtk_util {
 
+GdkPixbuf* GetPixbufFromIconID(const std::string& icon_id);
+GList* GetIconList();
+GList* ProductIconList();
+
 // Create a table of labeled controls, using proper spacing and alignment.
 // Arguments should be pairs of const char*, GtkWidget*, concluding with a
 // NULL.  The first argument is a vector in which to place all labels
--- a/chrome/common/pref_names.cc
+++ b/chrome/common/pref_names.cc
@@ -1212,6 +1212,9 @@
 // window to restore on startup.
 const char kBrowserWindowPlacement[] = "browser.window_placement";
 
+// Desktop Icon name
+const char kBrowserWindowIcon[] = "browser.window_icon";
+
 // A collection of position, size, and other data relating to the task
 // manager window to restore on startup.
 const char kTaskManagerWindowPlacement[] = "task_manager.window_placement";
--- a/chrome/common/pref_names.h
+++ b/chrome/common/pref_names.h
@@ -433,6 +433,7 @@
 extern const char kUninstallLastLaunchTimeSec[];
 extern const char kUninstallLastObservedRunTimeSec[];
 
+extern const char kBrowserWindowIcon[];
 extern const char kBrowserWindowPlacement[];
 extern const char kTaskManagerWindowPlacement[];
 extern const char kKeywordEditorWindowPlacement[];
--- a/chrome/common/chrome_switches.h
+++ b/chrome/common/chrome_switches.h
@@ -212,6 +212,7 @@
 extern const char kHostResolverRetryAttempts[];
 extern const char kHostResolverRules[];
 extern const char kHstsHosts[];
+extern const char kIconId[];
 extern const char kIgnoreGpuBlacklist[];
 extern const char kImport[];
 extern const char kImportFromFile[];
--- a/chrome/browser/ui/gtk/browser_titlebar.cc
+++ b/chrome/browser/ui/gtk/browser_titlebar.cc
@@ -602,14 +602,27 @@
         // top left of the custom frame, that will get updated when the
         // throbber is updated.
         Profile* profile = browser_window_->browser()->profile();
-        SkBitmap icon = browser_window_->browser()->GetCurrentPageIcon();
-        if (icon.empty()) {
-          gtk_util::SetWindowIcon(window_, profile);
-        } else {
-          GdkPixbuf* icon_pixbuf = gfx::GdkPixbufFromSkBitmap(icon);
-          gtk_util::SetWindowIcon(window_, profile, icon_pixbuf);
-          g_object_unref(icon_pixbuf);
+        // If we got an icon-id parameter (saved into prefs) and the
+        // current URL is a subpath of the app parameter use the icon-id
+        // icon
+
+        std::string current_host = browser_window_->browser()->GetURLForCurrentTab().host();
+
+        GList* icon_list = NULL;
+        GdkPixbuf* icon_pixbuf = NULL;
+        std::string icon_pref_key = browser_window_->browser()->GetWindowIconKey();
+        PrefService* prefs = browser_window_->browser()->profile()->GetPrefs();
+        FilePath prefs_icon_id = prefs->GetFilePath(icon_pref_key.c_str());
+        if (!prefs_icon_id.empty())
+          icon_pixbuf = gtk_util::GetPixbufFromIconID(prefs_icon_id.value());
+        if (icon_pixbuf) {
+          icon_list = g_list_append(icon_list, icon_pixbuf);
         }
+        else {
+          gtk_util::GetIconList();
+        }
+        gtk_window_set_icon_list(window_, icon_list);
+        g_list_free(icon_list);
         break;
       }
       case Browser::TYPE_TABBED: {
@@ -636,6 +649,7 @@
 
     // Note: we want to exclude the application popup/panel window.
     if ((browser_window_->browser()->is_app() &&
+         browser_window_->browser()->is_type_popup() &&
         !browser_window_->browser()->is_type_tabbed())) {
       SkBitmap icon = browser_window_->browser()->GetCurrentPageIcon();
       if (icon.empty()) {
--- a/chrome/browser/ui/extensions/application_launch.cc
+++ b/chrome/browser/ui/extensions/application_launch.cc
@@ -100,6 +100,8 @@
     return NULL;
   }
 
+  const FilePath::CharType kEmptyIcon[] = FILE_PATH_LITERAL("");
+  FilePath empty_icon(kEmptyIcon);
   switch (container) {
     case extension_misc::LAUNCH_NONE: {
       NOTREACHED();
@@ -122,7 +124,7 @@
     }
     case extension_misc::LAUNCH_WINDOW:
       tab = OpenApplicationWindow(profile, extension, container,
-                                  override_url, NULL);
+                                  override_url, empty_icon, NULL);
       break;
     case extension_misc::LAUNCH_TAB: {
       tab = OpenApplicationTab(profile, extension, override_url,
@@ -163,6 +165,7 @@
     const Extension* extension,
     extension_misc::LaunchContainer container,
     const GURL& url_input,
+    const FilePath& icon_id,
     Browser** app_browser) {
   DCHECK(!url_input.is_empty() || extension);
   GURL url = UrlForExtension(extension, url_input);
@@ -209,6 +212,12 @@
   if (app_browser)
     *app_browser = browser;
 
+  // save icon-id value for later use by UpdateWindowIcon
+  if (!icon_id.empty()) {
+      std::string icon_pref_key = browser->GetWindowIconKey();
+      profile->GetPrefs()->SetFilePath(icon_pref_key.c_str(), icon_id);
+  }
+
   TabContents* tab_contents =
       browser->AddSelectedTabWithURL(url, content::PAGE_TRANSITION_START_PAGE);
   WebContents* contents = tab_contents->web_contents();
@@ -231,6 +240,7 @@
 
 WebContents* OpenAppShortcutWindow(Profile* profile,
                                    const GURL& url,
+                                   const FilePath& icon_id,
                                    bool update_shortcut) {
   Browser* app_browser;
   WebContents* tab = OpenApplicationWindow(
@@ -238,6 +248,7 @@
       NULL,  // this is a URL app.  No extension.
       extension_misc::LAUNCH_WINDOW,
       url,
+      icon_id,
       &app_browser);
 
   if (!tab)
--- a/chrome/browser/ui/extensions/application_launch.h
+++ b/chrome/browser/ui/extensions/application_launch.h
@@ -55,6 +55,7 @@
     const extensions::Extension* extension,
     extension_misc::LaunchContainer container,
     const GURL& url,
+    const FilePath& icon_id,
     Browser** app_browser);
 
 // Open |url| in an app shortcut window.  If |update_shortcut| is true,
@@ -65,6 +66,7 @@
 // Browser::OpenApplicationWindow().
 content::WebContents* OpenAppShortcutWindow(Profile* profile,
                                             const GURL& url,
+                                            const FilePath& icon_id,
                                             bool update_shortcut);
 
 // Open an application for |extension| using |disposition|.  Returns NULL if
--- a/chrome/browser/ui/startup/startup_browser_creator_impl.cc
+++ b/chrome/browser/ui/startup/startup_browser_creator_impl.cc
@@ -504,10 +504,12 @@
         AppLauncherHandler::RecordAppLaunchType(
             extension_misc::APP_LAUNCH_CMD_LINE_APP_LEGACY);
       }
+      std::string icon_id(command_line_.GetSwitchValueASCII(switches:kIconId));
 
       WebContents* app_tab = application_launch::OpenAppShortcutWindow(
           profile,
           url,
+          FilePath(icon_id),
           true);  // Update app info.
 
       if (out_app_contents)
