--- a/chrome/browser/browser_main.cc
+++ b/chrome/browser/browser_main.cc
@@ -1267,12 +1267,10 @@
         !FirstRun::ProcessMasterPreferences(user_data_dir, &master_prefs);
     AddFirstRunNewTabs(&browser_init, master_prefs.new_tabs);
 
-    // If we are running in App mode, we do not want to show the importer
+    // Even if we are running in App mode, we do want to show the importer
     // (first run) UI.
     if (!first_run_ui_bypass &&
-        (parsed_command_line.HasSwitch(switches::kApp) ||
-         parsed_command_line.HasSwitch(switches::kAppId) ||
-         parsed_command_line.HasSwitch(switches::kNoFirstRun)))
+         parsed_command_line.HasSwitch(switches::kNoFirstRun))
       first_run_ui_bypass = true;
   }
 
