Index: src/chrome/browser/ui/window_sizer.cc
===================================================================
--- src.orig/chrome/browser/ui/window_sizer.cc	2010-03-16 20:59:03.000000000 +0100
+++ src/chrome/browser/ui/window_sizer.cc	2010-03-16 20:59:31.000000000 +0100
@@ -194,8 +194,8 @@
 
   // The default size is either some reasonably wide width, or if the work
   // area is narrower, then the work area width less some aesthetic padding.
-  int default_width = std::min(work_area.width() - 2 * kWindowTilePixels, 1050);
-  int default_height = work_area.height() - 2 * kWindowTilePixels;
+  int default_width = std::min(work_area.width() - 2 * kWindowTilePixels, 1024);
+  int default_height = std::min(work_area.height() - 2 * kWindowTilePixels, 576);
 
   // For wider aspect ratio displays at higher resolutions, we might size the
   // window narrower to allow two windows to easily be placed side-by-side.
