diff -Nur old/third_party/launchpad_translations/chromium_strings_ast.xtb new/third_party/launchpad_translations/chromium_strings_ast.xtb
--- old/third_party/launchpad_translations/chromium_strings_ast.xtb	2011-03-10 17:03:15.000000000 +0000
+++ new/third_party/launchpad_translations/chromium_strings_ast.xtb	2011-03-18 13:24:37.000000000 +0000
@@ -97,4 +97,12 @@
 <translation id="9191268552238695869">Un alministrador instaló Chromium nesti sistema y ta disponible pa tolos usuarios. Agora, la versión de Chromium a nivel de sistema va trocar la to instalación a nivel d'usuariu.</translation>
 <translation id="980185455267101462">Esti equipu yá tien una versión más reciente de Chromium Frame. Si'l software nun ta funcionando, desinstala Chromium Frame y descárgalu de nueves.</translation>
 <translation id="985602178874221306">Los autores de Chromium</translation>
+<translation id="205900032229966545">Intentasti llegar a &lt;strong&gt;<ph name="DOMAIN"/>&lt;/strong&gt;, pero'l sirvidor presentó un certificáu caducáu. Nun hai información disponible qu'indique si'l certificáu vióse comprometíu dende la so expiración. Esto implica que Chromium nun puede garantizar que teas comunicando con &lt;strong&gt;<ph name="DOMAIN2"/>&lt;/strong&gt; y non con un atacante. Nun tendríes de siguir.</translation>
+<translation id="2682584270762546522">Intentasti llegar a &lt;strong&gt;<ph name="DOMAIN"/>&lt;/strong&gt;, pero'l sirvidor presentó un certificáu caducáu. Nun hai información disponible qu'indique si se pue tener confianza nel certificáu. Chromium nun puede garantizar que teas comunicando con &lt;strong&gt;<ph name="DOMAIN2"/>&lt;/strong&gt; y non con un atacante. Deberíes asegurate de que'l reló y la estaya horaria del to equipu seyan les correutes. Si nun son correutes, configúrales bien, y anova esta páxina, Si son correutes, nun debieras siguir.</translation>
+<translation id="3849925841547750267">Llamentablemente, les tos preferencies de Mozilla Firefox nun van tar disponibles mientres qu'esi restolador se tea executando. Pa importar eses preferencies a Chromium, guarde'l so trabayu y zarre toles ventanes de Firefox. Después, calque en «Continuar».</translation>
+<translation id="4365115785552740256">Chromium ye una realidá gracies al proyeutu de códigu abiertu <ph name="BEGIN_LINK_CHROMIUM"/>Chromium<ph name="END_LINK_CHROMIUM"/> y a otros <ph name="BEGIN_LINK_OSS"/>programes de códigu abiertu<ph name="END_LINK_OSS"/>.</translation>
+<translation id="4387544446954190892">Chromium ye un restolador web qu'executa páxines web y aplicaciones a la velocidá de la lluz. Ye rápidu, estable y fácil d'usar. Navega pela rede más seguru cola proteición escontra software maliciosu y suplantación de personalidá que tien incorporaes Chromium.</translation>
+<translation id="4818993768077639880">Chromium yá ta instaláu y disponible pa tolos usuarios d'esti equipu. Si quies instalar Chromium a nivel d'usuariu, primero tienes de desinstalar la versión instalada a nivel de sistema por un alministrador.</translation>
+<translation id="4890456338028740446">Intentasti llegar a &lt;strong&gt;<ph name="DOMAIN"/>&lt;/strong&gt;, pero'l sirvidor presentó un certificáu emitíu por una entidá que nun tien la confianza del sistema operativu d'esti equipu. Esto pue significar que'l sirvidor xeneró les sos propies credenciales de seguridá, y Chromium nun pue confiar nestes no que cinca a la información sobre identidá, o un atacante pue tar intentando interceutar les tos comunicaciones. Nun debieras siguir &lt;strong&gt;especialmente&lt;/strong&gt; si nun teníes visto esti avisu nesti sitiu.</translation>
+<translation id="4931128279570440880">Nesti casu, la direición llistada nel certificáu nun correspuende cola direición del sitiu web al que'l to restolador intentó dir. Una posible razón ye que les tos comunicaciones tán interceutándose por un atacante que ta presentando un certificáu d'un sitiu web distintu, lo que podría causar una discordancia. Otra posible razón ye que'l sirvidor ta configuráu pa devolver el mesmu certificáu pa varios sitios web, incluyendo'l que tas intentando visitar, magar que'l certificáu nun ye válidu pa toos esos sitios. Chromium pue asegurar qu'aportesti a &lt;strong&gt;<ph name="DOMAIN2"/>&lt;/strong&gt;, pero nun pue verificar que seya'l mesmu sitiu &lt;strong&gt;<ph name="DOMAIN"/>&lt;/strong&gt; al que quies aportar. Si sigues, Chromium nun comprobará más discordancies de nomes. En xeneral, ye meyor nun siguir más allá d'esti puntu.</translation>
 </translationbundle>
\ No newline at end of file
diff -Nur old/third_party/launchpad_translations/chromium_strings_fi.xtb new/third_party/launchpad_translations/chromium_strings_fi.xtb
--- old/third_party/launchpad_translations/chromium_strings_fi.xtb	2011-03-10 17:03:15.000000000 +0000
+++ new/third_party/launchpad_translations/chromium_strings_fi.xtb	2011-03-18 13:24:37.000000000 +0000
@@ -75,4 +75,8 @@
 <translation id="7278516633989165382">Chromium Frame -kehys on päivitetty. Käynnistä selain uudelleeen. Chromen versio: <ph name="TODO_0001"/>, Chromium Frame -kehyksen versio: <ph name="TODO_0002"/></translation>
 <translation id="8046283042877435986">Tämä sivusto suosittelee Chromium Frame -kehystä (jo asennettu).</translation>
 <translation id="8971920611044657379">Chromium Frame -kehys</translation>
+<translation id="2632329825580776666">Asetustiedosto on korruptoitunut tai muuten virheellinen.\n\nChromium ei voi palauttaa asetuksiasi.</translation>
+<translation id="4222661289444749267">Chromium Frame -kehyksen päivitys</translation>
+<translation id="4330297532943865716">Järjestelmästä löytyi ristiriidan aiheuttava asennus Chromiumista tai Chromium Frame -kehyksestä. Poista se ensin ja yritä sitten asennusta uudelleen.</translation>
+<translation id="980185455267101462">Tällä tietokoneella on jo uudempi versio Chromium Frame -kehyksestä asennettuna. Jos se ei toimi kunnolla, poista Chromium Frame -kehyksen asennus ja lataa se uudelleen.</translation>
 </translationbundle>
\ No newline at end of file
diff -Nur old/third_party/launchpad_translations/chromium_strings_he.xtb new/third_party/launchpad_translations/chromium_strings_he.xtb
--- old/third_party/launchpad_translations/chromium_strings_he.xtb	2011-03-10 17:03:15.000000000 +0000
+++ new/third_party/launchpad_translations/chromium_strings_he.xtb	2011-03-18 13:24:37.000000000 +0000
@@ -106,4 +106,14 @@
 <translation id="985602178874221306">יוצרי Chromium</translation>
 <translation id="4330297532943865716">התקנה הגורמת להתנגשות של Chromium או מסגרת Chromium נמצאה במערכת. נא להסיר אותה ולנסות שוב.</translation>
 <translation id="8974095189086268230">Chromium OS זמינה בזכות <ph name="BEGIN_LINK_CROS_OSS"/>תכנה בקוד פתוח<ph name="END_LINK_CROS_OSS"/> נוספת.</translation>
+<translation id="1293235220023151515">התקנה סותרת של Chromium נמצאה במערכת. נא להסיר אותה ולנסות שוב.</translation>
+<translation id="1929939181775079593">Chromium אינו מגיב. אם להפעיל אותו מחדש כעת?</translation>
+<translation id="275588974610408078">אין אפשרות לדווח על קריסות ב־Chromium.</translation>
+<translation id="2927042866129133279">הארגומנטים שגויים. לא ניתן להתקין את מסגרת Chromium במצב הכנה מבלי להתקין גם את Chromium.</translation>
+<translation id="3096463977885168586">זכויות יוצרים © 2011-2006 יוצרי Chromium. כל הזכויות שמורות.</translation>
+<translation id="3190315855212034486">בום! Chromium קרס. האם להפעיל אותו מחדש כעת?</translation>
+<translation id="4077262827416206768">נא לסגור את כל החלונות של Chromium ולהפעיל אותו מחדש כדי שהשינויים ייכנסו לתוקף.</translation>
+<translation id="4649468518813464896">http://www.google.com/support/chrome/bin/answer.py?hl=[GRITLANGCODE]&amp;answer=161796</translation>
+<translation id="7278516633989165382">מסגרת Chromium עודכנה. נא להפעיל מחדש את הדפדפן. גרסת ה־Chromium:‏ <ph name="TODO_0001"/>, גרסת מסגרת ה־Chromium:‏ <ph name="TODO_0002"/></translation>
+<translation id="8046283042877435986">אתר זה ממליץ להשתמש במסגרת Chromium (כבר מותקנת).</translation>
 </translationbundle>
\ No newline at end of file
diff -Nur old/third_party/launchpad_translations/chromium_strings_nl.xtb new/third_party/launchpad_translations/chromium_strings_nl.xtb
--- old/third_party/launchpad_translations/chromium_strings_nl.xtb	2011-03-10 17:03:15.000000000 +0000
+++ new/third_party/launchpad_translations/chromium_strings_nl.xtb	2011-03-18 13:24:37.000000000 +0000
@@ -109,4 +109,11 @@
 <translation id="3096463977885168586">Copyright © 2006-2011 De auteurs van Chromium. Alle rechten voorbehouden.</translation>
 <translation id="4330297532943865716">Er werd een conflicterende installatie van Chromium of Chromium Frame op uw systeem gevonden. Verwijder deze versie van uw systeem en probeer opnieuw.</translation>
 <translation id="8046283042877435986">Deze website beveelt ‘Chromium Frame’ aan (Reeds geïnstalleerd).</translation>
+<translation id="1929939181775079593">Chromium reageert niet. Wilt u Chromium herstarten?</translation>
+<translation id="275588974610408078">Het doorgeven van crashes is niet mogelijk in Chromium.</translation>
+<translation id="2927042866129133279">Ongeldige opdracht. Chromium Frame kan niet in ready mode geïnstalleerd worden zonder ook Chromium te installeren.</translation>
+<translation id="3190315855212034486">Oeps! Chromium is gecrasht. Opnieuw starten?</translation>
+<translation id="4077262827416206768">U moet alle Chromium vensters sluiten en Chromium opnieuw opstarten voordat deze wijzigingen doorgevoerd kunnen worden.</translation>
+<translation id="4649468518813464896">http://www.google.com/support/chrome/bin/answer.py?hl=[GRITLANGCODE]&amp;answer=161796</translation>
+<translation id="7278516633989165382">Chromium Frame is geüpdatet, herstart uw browser. Chrome versie: <ph name="TODO_0001"/>, Chromium Frame versie: <ph name="TODO_0002"/></translation>
 </translationbundle>
\ No newline at end of file
diff -Nur old/third_party/launchpad_translations/chromium_strings_no.xtb new/third_party/launchpad_translations/chromium_strings_no.xtb
--- old/third_party/launchpad_translations/chromium_strings_no.xtb	2011-03-10 17:03:15.000000000 +0000
+++ new/third_party/launchpad_translations/chromium_strings_no.xtb	2011-03-18 13:24:37.000000000 +0000
@@ -43,4 +43,36 @@
 <translation id="8865765905101981392">Nettleser</translation>
 <translation id="911206726377975832">Slette dine brukerdata også?</translation>
 <translation id="985602178874221306">Chromium-utviklerne</translation>
+<translation id="1221340462641866827">Chromium OS støtter ikke oppstart av eksterne programmer for å kjøre <ph name="SCHEME"/>-koblinger. Den ønskede koblingen er  <ph name="PROTOLINK"/>.</translation>
+<translation id="1293235220023151515">En eksisterende installasjon av Chromium ble oppdaget på systemet. Vennligst avinstaller dette og prøv på nytt.</translation>
+<translation id="1929939181775079593">Chromium svarer ikke. Starte på nytt?</translation>
+<translation id="1967743265616885482">En modul med samme navn er i konflikt med Chromium.</translation>
+<translation id="2407806170074302369">Om Chromiumrammen...</translation>
+<translation id="2648074677641340862">En feil knyttet til operativsystemet oppsto under installasjonen. Vennligst last ned Chromium på nytt.</translation>
+<translation id="275588974610408078">Kræsjrapportering er ikke tilgjengelig i Chromium.</translation>
+<translation id="2843244238218937196">Chromium gjengiver</translation>
+<translation id="2927042866129133279">Ugyldige argumenter. Chromiumrammen kan ikke installeres i klar-modus uten også å installere Chromium.</translation>
+<translation id="3103660991484857065">Installasjonsprogrammet mislyktes å pakke ut filen. Vennligst last ned Chromium på nytt.</translation>
+<translation id="3190315855212034486">Aiai! Chromium har kræsjet. Start på nytt?</translation>
+<translation id="3509308970982693815">Vennligst lukk alle Chromiumvinduer og prøv på nytt.</translation>
+<translation id="3838208658251026018">Chromium importerer nå følgende elementer fra <ph name="BROWSER_COMPONENT"/>:</translation>
+<translation id="3849925841547750267">Dessverre er ikke innstillingene fra Mozilla Firefox tilgjengelig mens nettleseren kjører. Lagre arbeidet og lukk alle Firefoxvinduer for å importere disse til Chromium. Klikk deretter Fortsetter.</translation>
+<translation id="4077262827416206768">Lukk alle Chromiumvinduer og start Chromium på nytt for at endringene skal tre i kraft.</translation>
+<translation id="4222661289444749267">Chromiumramme-oppdatering</translation>
+<translation id="4285930937574705105">Installasjon milsyktes av ukjent årsak. Lukk Chromium hvis programmet fremdeles kjører.</translation>
+<translation id="4330585738697551178">Denne modulen går for å være i konflikt med Chromium.</translation>
+<translation id="4464530290860906238">Brukes ikke i Chromium. Plassholder for å holde ressurskart á jour.</translation>
+<translation id="4736009297872521264">Chromium oppfører seg ikke pent</translation>
+<translation id="5498103559159810756">Chromium bruker din standard søkemotor <ph name="PAGE_TITLE"/>. Ønsker du å beholde denne?</translation>
+<translation id="5559945513746540383">Chromium er ikke installert eller mislyktes å finne installasjonsmappen. Vennligst last ned Chromium på nytt.</translation>
+<translation id="5947389362804196214">Kan ikke lese innstillinger.\n\nDet kan hende at noen funksjoner ikke er tilgjengelige, og endringer i innstillinger vil ikke lagres.</translation>
+<translation id="6055895534982063517">Det finnes en ny versjon av Chromium, raskere enn noen gang.</translation>
+<translation id="6944967875980567883">Moduler lastet i Chromium</translation>
+<translation id="7123348595797445166">Prøv den ut (allerede installert)</translation>
+<translation id="7138853919861947730">Det kan hende Chromium bruker eksterne nettjenester for å forbedre nettopplevelsen.</translation>
+<translation id="7223968959479464213">Oppgavebehandling - Chromium</translation>
+<translation id="7278516633989165382">Chromium Frame har blitt oppdatert. Vennligst start nettleseren på nytt. Chrome versjonsnummer: <ph name="TODO_0001"/>, Chromium Frame versjonsnummer: <ph name="TODO_0002"/></translation>
+<translation id="7747138024166251722">Installasjonsprogrammet kunne ikke oppdrette en midlertidig mappe. Undersøk at det er nok ledig diskplass og at du har rettigheter til å installere programvare.</translation>
+<translation id="8046283042877435986">Dette nettsetdet anbefaler Chromium Frame (allerede installert).</translation>
+<translation id="9013262824292842194">Chromium må kjøre på Windows Vista eller Windows XP med SP2 eller høyere.</translation>
 </translationbundle>
\ No newline at end of file
